File: examples/deepnetwork.py
import os
import csv
import math
import time
import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader

# ==========================================
# 1. Model Definition (1:1 with Rust DeepNet)
# ==========================================
class DeepNet(nn.Module):
    def __init__(self, in_features: int = 9, hidden_dim: int = 64, out_features: int = 1):
        super().__init__()
        # In NeuralFlow, Linear has no bias and weights are initialized uniformly in [-1, 1]
        self.fc1 = nn.Linear(in_features, hidden_dim, bias=False)
        self.ln1 = nn.LayerNorm(hidden_dim, eps=1e-5)
        self.drop1 = nn.Dropout(p=0.2)

        self.fc2 = nn.Linear(hidden_dim, hidden_dim, bias=False)
        self.ln2 = nn.LayerNorm(hidden_dim, eps=1e-5)
        self.drop2 = nn.Dropout(p=0.2)

        self.fc3 = nn.Linear(hidden_dim, hidden_dim, bias=False)
        self.ln3 = nn.LayerNorm(hidden_dim, eps=1e-5)
        self.drop3 = nn.Dropout(p=0.2)

        self.fc4 = nn.Linear(hidden_dim, out_features, bias=False)
        self.relu = nn.ReLU()

        self._init_weights()

    def _init_weights(self):
        # Match Tensor::random in NeuralFlow: uniform(-1.0, 1.0)
        for m in [self.fc1, self.fc2, self.fc3, self.fc4]:
            nn.init.uniform_(m.weight, -1.0, 1.0)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.fc1(x)
        x = self.ln1(x)
        x = self.drop1(x)
        x = self.relu(x)

        x = self.fc2(x)
        x = self.ln2(x)
        x = self.drop2(x)
        x = self.relu(x)

        x = self.fc3(x)
        x = self.ln3(x)
        x = self.drop3(x)
        x = self.relu(x)

        x = self.fc4(x)
        return x


# ==========================================
# 2. Data Loading & Standard Scaler
# ==========================================
class StandardScaler:
    def __init__(self, eps: float = 1e-8):
        self.mean = None
        self.std = None
        self.eps = eps
        self.is_fitted = False

    def fit(self, data: np.ndarray):
        # Population mean and std (matching NeuralFlow's variance formula)
        self.mean = np.mean(data, axis=0)
        variance = np.mean(data ** 2, axis=0) - (self.mean ** 2)
        variance = np.maximum(variance, 0.0)
        self.std = np.sqrt(variance) + self.eps
        self.is_fitted = True

    def transform(self, data: np.ndarray) -> np.ndarray:
        assert self.is_fitted, "Scaler has not been fitted yet"
        return (data - self.mean) / self.std

    def inverse_transform(self, data: np.ndarray) -> np.ndarray:
        assert self.is_fitted, "Scaler has not been fitted yet"
        return (data * self.std) + self.mean


class HousingDataset(Dataset):
    def __init__(self, inputs: np.ndarray, targets: np.ndarray):
        self.inputs = torch.tensor(inputs, dtype=torch.float32)
        self.targets = torch.tensor(targets, dtype=torch.float32)

    def __len__(self):
        return len(self.inputs)

    def __getitem__(self, idx):
        return self.inputs[idx], self.targets[idx]


def load_housing_csv(csv_path: str, target_col_idx: int = 8, has_header: bool = True):
    inputs = []
    targets = []

    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)
        if has_header:
            next(reader, None)

        for row in reader:
            if not row:
                continue

            # In Rust CsvDataset, invalid floats (e.g. 'NEAR BAY') default to 0.0
            row_values = []
            for val in row:
                try:
                    row_values.append(float(val.strip()))
                except ValueError:
                    row_values.append(0.0)

            target = [row_values[target_col_idx]]
            input_feat = [v for idx, v in enumerate(row_values) if idx != target_col_idx]

            inputs.append(input_feat)
            targets.append(target)

    return np.array(inputs, dtype=np.float32), np.array(targets, dtype=np.float32)


# ==========================================
# 3. Metrics Evaluator (Matching Rust Evaluator)
# ==========================================
def calculate_metrics(y_pred: np.ndarray, y_true: np.ndarray):
    mse = np.mean((y_pred - y_true) ** 2)
    rmse = np.sqrt(mse)
    mae = np.mean(np.abs(y_pred - y_true))

    ss_tot = np.sum((y_true - np.mean(y_true)) ** 2)
    ss_res = np.sum((y_true - y_pred) ** 2)
    r2 = 1.0 - (ss_res / (ss_tot + 1e-8)) if ss_tot > 0 else 0.0

    return {"loss": mse, "r2": r2, "mae": mae, "rmse": rmse}


def evaluate(model: nn.Module, dataloader: DataLoader, criterion: nn.Module):
    model.eval()
    total_loss = 0.0
    batch_count = 0
    all_preds = []
    all_targets = []

    with torch.no_grad():
        for x_batch, y_batch in dataloader:
            preds = model(x_batch)
            loss = criterion(preds, y_batch)

            total_loss += loss.item()
            batch_count += 1

            all_preds.append(preds.cpu().numpy())
            all_targets.append(y_batch.cpu().numpy())

    all_preds = np.concatenate(all_preds, axis=0)
    all_targets = np.concatenate(all_targets, axis=0)

    metrics = calculate_metrics(all_preds, all_targets)
    metrics["loss"] = total_loss / batch_count if batch_count > 0 else 0.0
    return metrics


# ==========================================
# 4. Main Training Pipeline
# ==========================================
def main():
    in_features = 9
    hidden_dim = 64
    out_features = 1
    learning_rate = 5e-2
    batch_size = 256
    epochs = 50
    checkpoint_path = "best_deepnet_py.pt"

    # Set seed for reproducible comparison
    np.random.seed(42)
    torch.manual_seed(42)

    print("--- Loading Dataset ---")
    csv_path = os.path.join(os.path.dirname(__file__), "housing.csv")
    inputs, targets = load_housing_csv(csv_path, target_col_idx=8, has_header=True)
    total_samples = len(inputs)
    print(f"Loaded {total_samples} total samples.")

    # Standard Scaling
    x_scaler = StandardScaler()
    x_scaler.fit(inputs)
    inputs_scaled = x_scaler.transform(inputs)

    y_scaler = StandardScaler()
    y_scaler.fit(targets)
    targets_scaled = y_scaler.transform(targets)

    # Train / Val / Test split (70% / 15% / 15%)
    indices = np.random.permutation(total_samples)
    train_end = int(0.70 * total_samples)
    val_end = int(0.85 * total_samples)

    train_idx = indices[:train_end]
    val_idx = indices[train_end:val_end]
    test_idx = indices[val_end:]

    print(
        f"Dataset split: {len(train_idx)} train, {len(val_idx)} validation, {len(test_idx)} test samples."
    )

    train_dataset = HousingDataset(inputs_scaled[train_idx], targets_scaled[train_idx])
    val_dataset = HousingDataset(inputs_scaled[val_idx], targets_scaled[val_idx])
    test_dataset = HousingDataset(inputs_scaled[test_idx], targets_scaled[test_idx])

    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

    model = DeepNet(in_features, hidden_dim, out_features)
    criterion = nn.MSELoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

    best_val_loss = float("inf")

    print("\n--- Starting Training ---")
    start_time = time.time()

    for epoch in range(1, epochs + 1):
        model.train()
        total_train_loss = 0.0
        train_batches = 0

        for inputs_batch, targets_batch in train_loader:
            optimizer.zero_grad()
            pred = model(inputs_batch)
            loss = criterion(pred, targets_batch)
            loss.backward()
            optimizer.step()

            loss_val = loss.item()
            if not math.isnan(loss_val):
                total_train_loss += loss_val
                train_batches += 1

        avg_train_loss = total_train_loss / train_batches if train_batches > 0 else 0.0
        val_metrics = evaluate(model, val_loader, criterion)
        val_loss = val_metrics["loss"]

        # Checkpoint best model
        if val_loss < best_val_loss:
            print(f"Loss improved from {best_val_loss:.5f} to {val_loss:.5f}. Saving checkpoint to '{checkpoint_path}'")
            best_val_loss = val_loss
            torch.save(
                {
                    "epoch": epoch,
                    "model_state_dict": model.state_dict(),
                    "val_loss": val_loss,
                    "metrics": val_metrics,
                },
                checkpoint_path,
            )

        print(
            f"Epoch {epoch:2d}/{epochs} - Train MSE: {avg_train_loss:.4e} | "
            f"Val: Loss: {val_loss:.5f} | R2: {val_metrics['r2']:.4f} | "
            f"MAE: {val_metrics['mae']:.4f} | RMSE: {val_metrics['rmse']:.4f}"
        )

    elapsed_time = time.time() - start_time
    print(f"\nTraining completed in {elapsed_time:.2f}s")

    # ==========================================
    # 5. Testing with Best Checkpoint
    # ==========================================
    print("\n--- Loading Best Checkpoint for Testing ---")
    checkpoint = torch.load(checkpoint_path)
    model.load_state_dict(checkpoint["model_state_dict"])
    print(f"Loaded model from epoch {checkpoint['epoch']} with best val loss: {checkpoint['val_loss']:.4e}")

    test_metrics = evaluate(model, test_loader, criterion)
    print(
        f"Test Set Evaluation -> Loss: {test_metrics['loss']:.5f} | R2: {test_metrics['r2']:.4f} | "
        f"MAE: {test_metrics['mae']:.4f} | RMSE: {test_metrics['rmse']:.4f}"
    )

    # ==========================================
    # 6. Sample Inference
    # ==========================================
    print("\n--- Inference on Sample Test Data ---")
    model.eval()
    sample_count = min(5, len(test_dataset))
    with torch.no_grad():
        for i in range(sample_count):
            sample_inp, sample_tgt = test_dataset[i]
            pred_scaled = model(sample_inp.unsqueeze(0)).item()
            actual_scaled = sample_tgt.item()

            pred_val = y_scaler.inverse_transform(np.array([[pred_scaled]]))[0, 0]
            actual_val = y_scaler.inverse_transform(np.array([[actual_scaled]]))[0, 0]
            abs_err = abs(pred_val - actual_val)

            print(
                f"Sample #{i + 1}: Predicted = ${pred_val:,.2f} | "
                f"Actual = ${actual_val:,.2f} | Absolute Error = ${abs_err:,.2f}"
            )


if __name__ == "__main__":
    main()

File: examples/deepnetwork.rs
use NeuralFlow::prelude::*;
use std::collections::HashMap;

pub struct DeepNet {
    fc1: Linear,
    ln1: LayerNorm,
    drop1: Dropout,
    fc2: Linear,
    ln2: LayerNorm,
    drop2: Dropout,
    fc3: Linear,
    ln3: LayerNorm,
    drop3: Dropout,
    fc4: Linear,
}

impl DeepNet {
    pub fn new(in_dim: usize, hidden_dim: usize, out_dim: usize) -> Self {
        Self {
            fc1: Linear::new(in_dim, hidden_dim),
            ln1: LayerNorm::new(hidden_dim),
            drop1: Dropout::new(0.2),
            fc2: Linear::new(hidden_dim, hidden_dim),
            ln2: LayerNorm::new(hidden_dim),
            drop2: Dropout::new(0.2),
            fc3: Linear::new(hidden_dim, hidden_dim),
            ln3: LayerNorm::new(hidden_dim),
            drop3: Dropout::new(0.2),
            fc4: Linear::new(hidden_dim, out_dim),
        }
    }
}

impl Module for DeepNet {
    fn forward(&self, x: &Tensor) -> Tensor {
        let x = self.fc1.forward(&x);
        let x = self.ln1.forward(&x);
        let x = self.drop1.forward(&x);
        let x = x.relu();
        let x = self.fc2.forward(&x);
        let x = self.ln2.forward(&x);
        let x = self.drop2.forward(&x);
        let x = x.relu();
        let x = self.fc3.forward(&x);
        let x = self.ln3.forward(&x);
        let x = self.drop3.forward(&x);
        let x = x.relu();
        let x = self.fc4.forward(&x);
        x
    }

    fn parameters(&self) -> Vec<Tensor> {
        let mut p = self.fc1.parameters();
        p.extend(self.ln1.parameters());
        p.extend(self.drop1.parameters());
        p.extend(self.fc2.parameters());
        p.extend(self.ln2.parameters());
        p.extend(self.drop2.parameters());
        p.extend(self.fc3.parameters());
        p.extend(self.ln3.parameters());
        p.extend(self.drop3.parameters());
        p.extend(self.fc4.parameters());
        p
    }

    fn set_training(&self, mode: bool) {
        self.drop1.set_training(mode);
        self.drop2.set_training(mode);
        self.drop3.set_training(mode);
    }
}

pub fn main() -> Result<(), Box<dyn std::error::Error>> {
    let in_features = 9;
    let hidden_dim = 64;
    let out_features = 1;

    let model = DeepNet::new(in_features, hidden_dim, out_features);
    let optimizer = SGD::new(5e-2);

    println!("--- Loading Dataset ---");
    let dataset = CsvDataset::from_file("examples/housing.csv", 8, true)?;
    println!("Loaded {} total samples.", dataset.len());

    let mut scaler = StandardScaler::new();
    scaler.fit(&dataset);
    let mut target_scaler = StandardScaler::new();
    target_scaler.fit_targets(&dataset);
    let dataset = TransformedDataset::with_target_scaler(&dataset, &scaler, &target_scaler);

    let (train_data, val_data, test_data) =
        train_val_test_split(&dataset, 0.7, 0.15, 0.15, true);
    println!(
        "Dataset split: {} train, {} validation, {} test samples.",
        train_data.len(),
        val_data.len(),
        test_data.len(),
    );

    let batch_size = 256;
    let epochs = 50;
    let checkpoint_path = "best_deepnet.bin";

    let mut train_loader = Dataloader::new(&train_data, batch_size, true);
    let mut val_loader = Dataloader::new(&val_data, batch_size, false);
    let mut best_saver = BestModelSaver::new(checkpoint_path);

    println!("\n--- Starting Training ---");
    let start_time = std::time::Instant::now();

    for epoch in 1..=epochs {
        model.train();
        let mut total_train_loss = 0.0;
        let mut train_batches = 0;

        for (inputs, targets) in train_loader.iter_batches() {
            model.zero_grad();

            let pred = model.forward(&inputs);
            let loss = mse_loss(&pred, &targets);

            loss.backward();
            optimizer.step(&model.parameters());

            let loss_val = loss.0.borrow().data[0];
            if !loss_val.is_nan() {
                total_train_loss += loss_val;
                train_batches += 1;
            }
        }

        let avg_train_loss = if train_batches > 0 {
            total_train_loss / train_batches as f32
        } else {
            0.0
        };

        // 1. Evaluation using Evaluator
        let val_report = Evaluator::evaluate(&model, &mut val_loader, mse_loss);

        // 2. Save best checkpoint
        let mut metrics_map = HashMap::new();
        metrics_map.insert("mae".to_string(), val_report.mae);
        metrics_map.insert("rmse".to_string(), val_report.rmse);
        metrics_map.insert("r2".to_string(), val_report.r2_score);
        best_saver.check_and_save(&model, epoch, val_report.loss, metrics_map);

        println!(
            "Epoch {:2}/{} - Train MSE: {:.4e} | Val: {}",
            epoch, epochs, avg_train_loss, val_report
        );
    }

    let elapsed = start_time.elapsed();
    println!("\nTraining completed in {:.2}s", elapsed.as_secs_f32());

    // 3. Load the best saved checkpoint
    println!("\n--- Loading Best Checkpoint for Testing ---");
    let mut loaded_model = DeepNet::new(in_features, hidden_dim, out_features);
    let checkpoint = ModelCheckpoint::load(&mut loaded_model, checkpoint_path)?;
    println!(
        "Loaded model from epoch {} with best val loss: {:.4e}",
        checkpoint.epoch, checkpoint.val_loss
    );

    let mut test_loader = Dataloader::new(&test_data, batch_size, false);
    let test_report = Evaluator::evaluate(&loaded_model, &mut test_loader, mse_loss);
    println!("Test Set Evaluation -> {}", test_report);

    // 4. Run sample inference
    println!("\n--- Inference on Sample Test Data ---");
    let sample_count = 5.min(test_data.len());
    for i in 0..sample_count {
        let (sample_inputs, sample_targets) = test_data.get(i);
        let input_tensor = Tensor::new(sample_inputs.clone(), (1, in_features));
        let pred_tensor = loaded_model.forward(&input_tensor);

        let pred_scaled = pred_tensor.0.borrow().data[0];
        let actual_scaled = sample_targets[0];
        let pred_val = target_scaler.inverse_transform_sample(&[pred_scaled])[0];
        let actual_val = target_scaler.inverse_transform_sample(&[actual_scaled])[0];
        let error = (pred_val - actual_val).abs();

        println!(
            "Sample #{}: Predicted = ${:.2} | Actual = ${:.2} | Absolute Error = ${:.2}",
            i + 1,
            pred_val,
            actual_val,
            error
        );
    }

    Ok(())
}
